src/lib/operators/join_index.cpp

```diff
@@ -311,30 +376,72 @@ void JoinIndex::_fallback_nested_loop(const ChunkID index_chunk_id, const bool t
   join_index_performance_data.chunks_scanned_without_index++;
 }
 
-// join loop that joins two segments of two columns using an iterator for the probe side,
-// and an index for the index side
+// Join loop that joins two segments of two columns using an iterator for the probe side,
+// and a table index for the index side.
 template <typename ProbeIterator>
-void JoinIndex::_data_join_two_segments_using_index(ProbeIterator probe_iter, ProbeIterator probe_end,
-                                                    const ChunkID probe_chunk_id, const ChunkID index_chunk_id,
-                                                    const std::shared_ptr<AbstractIndex>& index) {
+void JoinIndex::_data_join_two_segments_using_table_index(ProbeIterator probe_iter, ProbeIterator probe_end,
+                                                          const ChunkID probe_chunk_id,
+                                                          const std::shared_ptr<AbstractTableIndex>& table_index) {
   for (; probe_iter != probe_end; ++probe_iter) {
     const auto probe_side_position = *probe_iter;
-    const auto index_ranges = _index_ranges_for_value(probe_side_position, index);
+
+    auto append_matches = [&probe_side_position, &probe_chunk_id, this](auto index_begin, auto index_end) {
+      _append_matches_table_index(index_begin, index_end, probe_side_position.chunk_offset(), probe_chunk_id);
+    };
+
+    // AntiNullAsTrue is the only join mode in which comparisons with null-values are evaluated as "true".
+    // If the probe side value is null or at least one null value exists in the indexed join segment, the probe value
+    // has a match.
+    if (_mode == JoinMode::AntiNullAsTrue) {
+      if (probe_side_position.is_null() || table_index->indexed_null_values()) {
+        table_index->access_values_with_iterators(append_matches);
+        table_index->access_null_values_with_iterators(append_matches);
+      }
```

Comment:
The nesting level can be reduced by (1) using a guard condition (see `_chunk_index_ranges_for_value`), (2) merging the conditions in lines 395 and 396.